projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/control/LLVMLoopDispatchNode.java

```diff
@@ -0,0 +1,265 @@
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.oracle.truffle.llvm.nodes.control;
+
+import com.oracle.truffle.api.CompilerAsserts;
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.CompilerDirectives.CompilationFinal;
+import com.oracle.truffle.api.frame.FrameSlot;
+import com.oracle.truffle.api.frame.FrameSlotTypeException;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.instrumentation.StandardTags;
+import com.oracle.truffle.api.instrumentation.Tag;
+import com.oracle.truffle.api.nodes.ExplodeLoop;
+import com.oracle.truffle.api.nodes.ExplodeLoop.LoopExplosionKind;
+import com.oracle.truffle.llvm.nodes.base.LLVMBasicBlockNode;
+import com.oracle.truffle.llvm.nodes.base.LLVMFrameNullerUtil;
+import com.oracle.truffle.llvm.nodes.func.LLVMInvokeNode;
+import com.oracle.truffle.llvm.runtime.except.LLVMUserException;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMControlFlowNode;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMExpressionNode;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMStatementNode;
+
+public final class LLVMLoopDispatchNode extends LLVMExpressionNode {
+
+    private final FrameSlot exceptionValueSlot;
+    private final int headerId;
+    @Children private final LLVMStatementNode[] bodyNodes;
+    @CompilationFinal private final Integer[] indexMapping;
+    @CompilationFinal(dimensions = 2) private final FrameSlot[][] beforeBlockNuller;
+    @CompilationFinal(dimensions = 2) private final FrameSlot[][] afterBlockNuller;
+    @CompilationFinal private final Integer[] loopSuccessors;
+    @CompilationFinal private final FrameSlot successorSlot;
+
+    public LLVMLoopDispatchNode(FrameSlot exceptionValueSlot, LLVMStatementNode[] bodyNodes, FrameSlot[][] beforeBlockNuller, FrameSlot[][] afterBlockNuller, int headerId, Integer[] indexMapping, Integer[] successors, FrameSlot successorSlot) {
+        this.exceptionValueSlot = exceptionValueSlot;
+        this.bodyNodes = bodyNodes;
+        this.beforeBlockNuller = beforeBlockNuller;
+        this.afterBlockNuller = afterBlockNuller;
+        this.indexMapping = indexMapping;
+        this.headerId = headerId;
+        this.loopSuccessors = successors;
+        this.successorSlot = successorSlot;
+    }
+
+    @ExplodeLoop
+    private boolean isInLoop(int bci) {
+        for(int i : loopSuccessors) {
+            if(i == bci)
+                return false;
+        }
+        return true;
+    }
+
+    @Override
+    @ExplodeLoop(kind = LoopExplosionKind.MERGE_EXPLODE)
+    public Object executeGeneric(VirtualFrame frame) {
+        boolean ret = false; // helper variable to not return at first encounter of headerId
```

Comment:
something like firstIteration would be more readable and would make the comment unnecessary